### PR TITLE
Enables signposts for core

### PR DESCRIPTION
**Pull request by trooper2013**
This PR enables SFUserAccountManger & SFRestAPI for Instrumentation.  The approach taken avoids the addition of signpost related code into our main code line. Maintains a clean separation of concerns. 

Note: Signposts feature is only available in iOS12 and above. There will be a very small performance impact of this approach. Ive tried to keep the instrumentation code as light as possible.     

1. Adds 2 Categories which swizzle out and wrap some functions for SFUserAccountManager and SFRestAPI.

2. Add macros to wrap the os_signpost functions to avoid for ugly code branching iOS 12 related.

3. Added a SIGNPOST_ENABLED compile var which is enabled in DEBUG by default. 

The next PR will add similar functionality to SFSmartStore and SFSmartSync.

### CODE REVIEW

File: libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -848,6 +848,7 @@
 				MTL_ENABLE_DEBUG_INFO = YES;
 				ONLY_ACTIVE_ARCH = YES;
 				SDKROOT = iphoneos;
+				SIGNPOST_ENABLED = 1;
```
Comment (wmathurin) [inline libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj]: Why do we have it in the xcconfig and in the project file?

File: configuration/Configuration-Debug.xcconfig [hunk 2]
```diff
@@ -11,4 +11,5 @@ COPY_PHASE_STRIP = NO
 
 GCC_OPTIMIZATION_LEVEL = 0
 GCC_PREPROCESSOR_DEFINITIONS = DEBUG=1 $(inherited)
+SIGNPOST_ENABLED = 1
```
Comment (wmathurin) [inline configuration/Configuration-Debug.xcconfig]: Shouldn't hat move into the previous line to be available to the preprocessor ?
```
GCC_PREPROCESSOR_DEFINITIONS = DEBUG=1 SIGMPOST_ENABLED=1 $(inherited)
```

File: configuration/Configuration-Release.xcconfig [hunk 3]
```diff
@@ -11,4 +11,5 @@ COPY_PHASE_STRIP = YES
 
 ENABLE_NS_ASSERTIONS = NO
 MTL_ENABLE_DEBUG_INFO = NO
+SIGNPOST_ENABLED = 0
```
Comment (wmathurin) [inline configuration/Configuration-Release.xcconfig]: Not defined or = 0 is the same thing, so we can probably get rid of that line.

File: libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj (hunk 1 above)
Comment (wmathurin) [reply libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj]: I don't think you need that one if you have it in the GCC_PREPROCESSOR_DEFINITIONS through the xcconfig file.

File: configuration/Configuration-Debug.xcconfig (hunk 2 above)
Comment (bhariharan) [reply configuration/Configuration-Debug.xcconfig]: +1

File: libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj (hunk 1 above)
Comment (bhariharan) [reply libs/SalesforceAnalytics/SalesforceAnalytics.xcodeproj/project.pbxproj]: Agreed - let's just keep it in the `xcconfig` file if possible.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Instrumentation/SFRestAPI+Instrumentation.m [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+@interface SFRestDelegateWrapperWithIntrumentation<SFRestDelegate>: NSObject
+- (instancetype)initWithDelegate:(id<SFRestDelegate>) delegate signpost:(os_signpost_id_t)signpostId logger:(os_log_t) logger;
+@property (weak,nonatomic,readonly) id<SFRestDelegate> delegate;
+@property (nonatomic,readonly) os_signpost_id_t signpostId;
+@property (nonatomic,readonly) os_log_t logger;
+
++(id<SFRestDelegate>)wrapperWith:delegate signpost:(os_signpost_id_t)signpostId logger:(os_log_t) logger;
+@end
+
+@implementation SFRestDelegateWrapperWithIntrumentation
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Instrumentation/SFRestAPI+Instrumentation.m]: Typo: should be `Instrumentation` (missing `s`).

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Instrumentation/SFRestAPI+Instrumentation.m [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+#import "SFRestAPI+Instrumentation.h"
+#import "SalesforceSDKConstants.h"
+#import "SFSDKInstrumentationHelper.h"
+#import <objc/runtime.h>
+#import <os/log.h>
+#import <os/signpost.h>
+#import "SFSDKCoreLogger.h"
+
+@interface SFRestDelegateWrapperWithIntrumentation<SFRestDelegate>: NSObject
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Instrumentation/SFRestAPI+Instrumentation.m]: Typo: should be `Instrumentation` (missing `s`).

**Pull request merged**